runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java

```diff
@@ -236,14 +236,26 @@ public long getConflictSetAndCommit(ConflictSetInfo conflictSet) {
         }
 
         // If the write set is empty, this is a read-only transaction.
-        // Return the max address of all accessed streams, this will provide a safe token of the
+        // If the transaction has only read non-monotonic objects, then return the
+        // max address of all accessed streams, this will provide a safe token of the
         // transaction's snapshot. Notice that, providing the snapshot of the tx (vs. max address)
         // can lead to data loss, as a sequencer reboot might incur in sequence regression.
         // This timestamp is aimed to provide clients with a secure point for delta/streaming
         // subscription, the later could lead to data loss scenarios.
+        // If the transaction has read monotonic objects, we instead return the min address
+        // of all accessed streams. Although this avoids data loss, clients subscribing at
+        // this point for delta/streaming may observe duplicate data.
         if (getWriteSetInfo().getWriteSet().getEntryMap().isEmpty()) {
             log.trace("Commit[{}] Read-only commit (no write)", this);
-            return getMaxAddressRead();
+
+            final boolean hasReadMonotonicObject = knownStreamsPosition.keySet()
+                    .stream().anyMatch(proxy -> proxy.getUnderlyingObject().isMonotonicObject());
```

Comment:
One other alternative would be to define an additional flag `accessedMonotonicObjects` in `AbstractTransactionalContext`. You would update it on every access. This way you don't need to pass the entire proxy. The downside is that you would be introducing additional state.

Reply:
Thanks - I have added this flag. It's worth noting that I still pass the proxy to `updateKnownStreamPosition` in order to leverage `isMonotonicObject` to determine which precondition check to perform and to update the newly added flag. However, only the streamId is stored in the `knownStreamsPosition` map.